### PR TITLE
support git token provider kind in the supervisor

### PR DESCRIPTION
**Pull request by akosyakov**
- [x] /werft https

#### What it does

- it should be used as a fallback for now for other IDEs, i.e. Code

#### How to test

- start a workspace: https://ak-git-token-in-supervisor.staging.gitpod-dev.com/#https://github.com/gitpod-io/go-gin-app
- checkout a new branch and try to push from the cli and the SCM view
- switch to Code, test again
- in Code check that links to home, workspaces and access control work properly
- stopping a workspace from Code should work as well

### CODE REVIEW

File: components/gitpod-cli/pkg/theialib/service.go [hunk 1]
```diff
@@ -88,6 +88,13 @@ func (service *HTTPTheiaService) sendRequest(req request) ([]byte, error) {
 	httpreq.Header.Set("Content-Type", "application/json")
 	httpreq.Header["X-AuthToken"] = []string{service.Token}
 	resp, err := service.ideClient.Do(httpreq)
+	if resp.StatusCode == 404 && req.Method == "getGitToken" {
+		supervisorAddr := os.Getenv("SUPERVISOR_ADDR")
+		if supervisorAddr == "" {
+			supervisorAddr = "localhost:22999"
+		}
+		resp, err = http.Get(fmt.Sprintf("http://%s/_supervisor/v1/token/git/%s/*", supervisorAddr, req.Params.(GetGitTokenRequest).Host))
```
Comment (akosyakov) [inline components/gitpod-cli/pkg/theialib/service.go:96]: later `scope` will be replaced from `*` with `{repo}:{command}` to support token validations

File: components/supervisor/pkg/supervisor/git.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+// NewGitTokenProvider creates a new instance of gitTokenProvider
+func NewGitTokenProvider(gitpodAPI gitpod.APIInterface) *GitTokenProvider {
+	return &GitTokenProvider{
+		gitpodAPI: gitpodAPI,
+	}
+}
+
+// GetToken resolves a token from a git hosting service
+func (p *GitTokenProvider) GetToken(ctx context.Context, req *api.GetTokenRequest) (tkn *Token, err error) {
+	token, err := p.gitpodAPI.GetToken(ctx, &gitpod.GetTokenSearchOptions{
```
Comment (akosyakov) [inline components/supervisor/pkg/supervisor/git.go:29]: What if gitpod api is nil?

Comment (svenefftinge) [reply components/supervisor/pkg/supervisor/git.go:29]: Seems like in that case the whole supervisor wouldn't work properly and it is ok to fail. I don't see how that could happen though.

**Comment by svenefftinge** [comment]
Thanks Anton! I think this is a necessary change to make the vscode variant useful at all. I like that it doesn't affect the existing and working theia approach but uses supervisor only through a fallback on 404.
We need to be aware however, that this changes makes it possible to get a token even if the original user is not connected through a web frontend. I think that is ok and we need to go into this direction anyway. I.e. we need to reconsider our collaborative mode and make it more restrictive by default (i.e. guest users can not run any processes).

**Review by svenefftinge — APPROVED** [review]
LGTM. I have tested the changes with both Theia and Code. Also read through the code.

**Comment by svenefftinge** [comment]
@csweichel please check as well 🙏

File: components/supervisor/pkg/supervisor/git.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+		Host: req.Host,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if token.Value == "" {
+		// token was not found, thus we return just a dummy to satisfy the git protocol
+		return &Token{
+			User:  "oauth2",
```
Comment (akosyakov) [inline components/supervisor/pkg/supervisor/git.go:38]: it is not enough to return it here, credential helper get should never fail, but return meaningful output, checked with @AlexTugarev

File: components/supervisor/pkg/supervisor/git.go (hunk 2 above)
Comment (akosyakov) [reply components/supervisor/pkg/supervisor/git.go:32]: In other cases we have guards.

File: components/supervisor/pkg/supervisor/git.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+import (
+	"context"
+	"time"
+
+	"github.com/gitpod-io/gitpod/supervisor/api"
+	"github.com/gitpod-io/gitpod/supervisor/pkg/gitpod"
+)
+
+// GitTokenProvider provides tokens from git hosting services
```
Comment (csweichel) [inline components/supervisor/pkg/supervisor/git.go:15]:
```
// GitTokenProvider provides tokens for Git hosting services by asking
// the Gitpod server.
```

File: components/supervisor/pkg/supervisor/services_test.go [hunk 5]
```diff
@@ -172,11 +196,12 @@ func TestInMemoryTokenServiceGetToken(t *testing.T) {
 		{
 			Desc: "token provider",
 			Req: &api.GetTokenRequest{
+				Kind:  defaultKind,
```
Comment (csweichel) [inline components/supervisor/pkg/supervisor/services_test.go:199]: could you add a test or two that tests the separation along the token `kind`?

File: components/supervisor/pkg/supervisor/supervisor.go [hunk 6]
```diff
@@ -210,8 +211,10 @@ func createGitpodService(cfg *Config, tknsrv api.TokenServiceServer) *gitpod.API
 		return nil
 	}
 	tknres, err := tknsrv.GetToken(context.Background(), &api.GetTokenRequest{
+		Kind: "gitpod",
```
Comment (csweichel) [inline components/supervisor/pkg/supervisor/supervisor.go:214]: maybe we should have a few constants in `supervisor-api` for common token kinds.
E.g.

```
const (
    // KindGitpod marks tokens that provide access to the Gitpod server API
    KindGitpod = "gitpod"
    
    // KindGit marks any kind of Git access token
    KindGit = "git"
)
```

and something like that on the TS side if that makes sense (some uses that).

**Review by csweichel — COMMENTED** [review]
had a look at the code ... now testing

**Review by csweichel — APPROVED** [review]
Works like a charm

File: components/gitpod-cli/cmd/credential-helper.go [hunk 7]
```diff
@@ -45,7 +58,8 @@ var credentialHelper = &cobra.Command{
 
 		service, err := theialib.NewServiceFromEnv()
 		if err != nil {
-			log.Fatal(err)
+			log.Print(err)
```
[the remark this answers is not in the document]
Comment (akosyakov) [reply components/gitpod-cli/cmd/credential-helper.go:61]: I don't know how to add a dep to gitpod-io/common here. I tried to change BUILD.yaml and go.mod, but go is not happy.

[the remark this answers is not in the document]
Comment (csweichel) [reply components/gitpod-cli/cmd/credential-helper.go:61]: Sorry, that's my bad. I didn't think about this being gp-cli, hence wasn't specific enough. I doubt it's worth pulling in common-go just for logs.

We could just use logrus instead:
```
impport (
  log "github.com/sirupsen"
)
```

**Comment by svenefftinge** [comment]
just stumbled over https://github.com/gitpod-io/gitpod/issues/2368

**Comment by gtsiolis** [comment]
I'm late to the party but this looks good! 🥕 

Thanks for fixing https://github.com/gitpod-io/gitpod/issues/2256 here @akosyakov! 🙏 

> ... stopping a workspace from Code should work as well

It took me a while to realize I can only stop the workspace through the command palette in Code editor. /cc @akosyakov

**Comment by akosyakov** [comment]
@gtsiolis You can also stop it by clicking Gitpod in the status bar.

**Comment by gtsiolis** [comment]
> You can also stop it by clicking Gitpod in the status bar.

Aha! I see this is an existing interaction pattern for the Code editor which is great. Thanks @akosyakov! 🏀

**Pull request merged**